### PR TITLE
Wip xio

### PR DESCRIPTION
**Pull request by mattbenjamin**
This is a pre-pull request, designed to give folks a chance to do early review of the current XioMessenger integration for Hammer.

### CODE REVIEW

**Comment by mattbenjamin** [comment]
This rebase hasn't seen full testing by Vu, but the input branch passed.  I figured better to stage the other changes.

**Comment by mattbenjamin** [comment]
This is rebase of changes I pushed yesterday, fixing autotools build issues.  (Previous branch available at wip-xio-0101.)

**Comment by mattbenjamin** [comment]
The buildbot issue seems spurious--I noticed the same error when I first renamed rbd-fuse.c, but cleaning up the build tree fixed it...

**Comment by liewegas** [comment]
Please drop these two patches:

- librados: clear g_ceph_context on shutdown (definitely not correct)
- Add GENERIC_SERVER entity name type. (just re-use mon or some other type, it's cosmetic for a test)

I think with that this is pretty close.  I'm still not a big fan of the #defines in Message.h for the tracing stuff, but it's not a big deal.  Maybe a comment explaining what it is and which field gets them...

**Comment by mattbenjamin** [comment]
I think the automated build is confused.  It needs to start fresh.

**Comment by mattbenjamin** [comment]
entity_name_t:  with some guidelines for validation, we can send another changeset in future.  I'm not sure just creating the GENERIC name is that useful, by itself.

**Comment by liewegas** [comment]
yeah given that it's only used for a test i don't think it's worth any 
time investment :)

**Comment by mattbenjamin** [comment]
I'm uncertain why you wouldn't want to make simple changes that make it easier to write new programs using Messenger, as you would in ONC, DCE, or whatever.  It didn't take much time investment to add an unclaimed name, it's taken more time to discuss it than it did to write.  The lack of it was, however, one of the first things that struck me when I started working on XioMessenger.  Maybe I'm just too used to RPC systems.

**Comment by yuyuyu101** [comment]
It's looks like that we still need some pre-cleanup commits to make it simple. Such as crc and tracepoint jobs.

**Comment by mattbenjamin** [comment]
I don't follow.

**Comment by dachary** [comment]
@mattbenjamin 
> I think the automated build is confused. It needs to start fresh.

the bot has been fixed, thanks for your patience.

**Comment by mattbenjamin** [comment]
No, thank you.  I just recognized the error as one I'd seen in test builds, and resetting my build had fixed it.

Matt

----- "Loic Dachary" <notifications@github.com> wrote:

> @mattbenjamin
> 
> 
> 
> I think the automated build is confused. It needs to start fresh.
> 
> the bot has been fixed, thanks for your patience.
> 
> —
> Reply to this email directly or view it on GitHub .

-- 
Matt Benjamin
CohortFS, LLC.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://cohortfs.com

tel.  734-761-4689 
fax.  734-769-8938 
cel.  734-216-5309

**Comment by liewegas** [comment]
couple minor issues but no blockers i see!  will include this in the test run

**Comment by mattbenjamin** [comment]
thank you, sage--should I let it run, push another commit, or rebase?

**Comment by liewegas** [comment]
On Fri, 9 Jan 2015, Matt Benjamin wrote:
> thank you, sage--should I let it run, push another commit, or rebase?

just one thing that fails the make check:

 test/ceph-helpers.sh:    test $(get_config mon a ms_nocrc) = false || return 1

can you fix that up?

With that change and making sure the Makefile.am's have the nen headers I 
think we're good! thanks-

**Comment by mattbenjamin** [comment]
yeah, sure

----- "Sage Weil" <notifications@github.com> wrote:

> On Fri, 9 Jan 2015, Matt Benjamin wrote:
> > thank you, sage--should I let it run, push another commit, or
> rebase?
> 
> just one thing that fails the make check:
> 
> test/ceph-helpers.sh: test $(get_config mon a ms_nocrc) = false ||
> return 1
> 
> can you fix that up?
> 
> With that change and making sure the Makefile.am's have the nen
> headers I
> think we're good! thanks-
> 
> —
> Reply to this email directly or view it on GitHub .

-- 
Matt Benjamin
CohortFS, LLC.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://cohortfs.com

tel.  734-761-4689 
fax.  734-769-8938 
cel.  734-216-5309

**Comment by mattbenjamin** [comment]
Ok, think I got that.

**Comment by liewegas** [comment]
@mattbenjamin Can you track down where the link error loic-bot sees above is coming from?  I'm seeing the same thing on my local box.

**Comment by mattbenjamin** [comment]
I am not seeing this our dev boxes, which are Ubuntu 14.04 and Debian unstable.  What are you and or Loic-bot running on?

**Comment by liewegas** [comment]
loic-bot is fedora 20, my box is ubuntu 12.04

**Comment by mattbenjamin** [comment]
Thanks.  I assume the fix is to make boost-regex linkage explicit.  I'm not sure why it's needed.

**Comment by dachary** [comment]
@mattbenjamin loic-bot is actually 14.04 for the https://github.com/ceph/ceph/pull/3224#issuecomment-69470738 run (which is running in a container on a fedora 20 bare metal as mentionned by @liewegas ). Starting today it's running in a centos 7 container (and will display the information in the message to avoid further confusion).

**Comment by mattbenjamin** [comment]
Thanks.  I haven't been able to reproduce yet.  But I'm sure it's my fault :)
I decided to build on a new Fedora 21 vm, with XFS rootfs.  After autogen, I've been doing (on all the builds) just "./configure --prefix=/opt/wip-xio";  Make builds to completion.  In all my builds, just doing "make check" from top srcdir fails before really doing anything (and I'm not sure why).  But "make check" from top_srcdir/src succeeds, mostly (I just saw a unittest_arch fail on F21, and "make check" appears to be stuck on unittest_chain_xattr, hopefully it will finish up...) .  Is this behavior wrt the src directory what I should be observing?  Am I using the right configure invocation?

**Comment by mattbenjamin** [comment]
It turned out that unittest_flatindex was hung, and had logged that it couldn't figure out the number of running threads.  That must be transient, because running "make check" again did better.  The unittest_arch fails consistently, but that seems to be related to my host CPU?  Most things worked once xmlstarlet was installed.  test/ceph_objectstore_tool.py failed reporting that OSD held the store locked (deterministic?).  cephtool-test-mon.sh reported FAIL, but I don't see an error condition in its log file.  Everything else reported PASS.  Not that this likely matters, I guess.  I'm just trying to provide context about what I might be doing wrong.

Obviously, I don't see the boost-regex link issue.  I should try a Centos7 vm next, I guess.

**Comment by liewegas** [comment]
FWIW it's a build time issue, not make check, so if yo uget that far 
you're past it.  I generally run configure via 
./do_autogen.sh -d 3 -T'.

I added you to the ceph team on github.  If you push your wip-xio to the 
main ceph.git the builders will have a go at it and you can see the 
results at

http://ceph.com/gitbuilder.cgi

Clicking the red build and following the error link will show the full 
output.  Hopefully that will help?  You can also grab a random 12.04 
machien in the lab (I forget if you have access yet?) and try building 
there.

**Comment by mattbenjamin** [comment]
Thanks, Sage.  I'll start getting up to speed with it.

**Comment by mattbenjamin** [comment]
I got 403 trying to push wip-xio to ceph/ceph.git--which should have worked per above?

**Comment by liewegas** [comment]
you may have to 'git remote set-url <name> git@github.com:ceph/ceph' if it 
was a git:// url before?

**Comment by mattbenjamin** [comment]
sorry, I should have said, I did retest with git+ssh

----- "Sage Weil" <notifications@github.com> wrote:

> you may have to 'git remote set-url <name> git@github.com:ceph/ceph'
> if it
> was a git:// url before?
> 
> —
> Reply to this email directly or view it on GitHub .

-- 
Matt Benjamin
CohortFS, LLC.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://cohortfs.com

tel.  734-761-4689 
fax.  734-769-8938 
cel.  734-216-5309

**Comment by liewegas** [comment]
oh, you have to accept the github invitation to join the ceph test... 
probably at github.com/mattbenjamin somewhere?

**Comment by mattbenjamin** [comment]
ah, that's better

**Comment by mattbenjamin** [comment]
There were still some files missing from noinst_HEADERS, but with that fixed, the issue is just that we should require rather than "notice" linking with boost_regex, which isn't header-only.  What's the process for adding such a dependency?

**Comment by mattbenjamin** [comment]
Sorry (per IRC scrollback), I wasn't clear.  I had added a check for boost-regex libs to configure.ac, but copied the logic that was already there to find boost-system.  That was a mistake, because that check warns but does not stop the configure step if a library is not found.

**Comment by mattbenjamin** [comment]
Any update?

**Comment by dachary** [comment]
@mattbenjamin if you rebase + repush that will trigger the make check bot

**Comment by liewegas** [comment]
i'll try this locally and hopefully clean up the last issues.

@mattbenjamin, can we make the new dependency only necessary when XIO 
is enabled?  is xio the only user of address_helper?

**Comment by mattbenjamin** [comment]
atm yes, so I could move it, sure

----- "Sage Weil" <notifications@github.com> wrote:

> i'll try this locally and hopefully clean up the last issues.
> 
> @mattbenjamin, can we make the new dependency only necessary when XIO
> is enabled? is xio the only user of address_helper?
> 
> —
> Reply to this email directly or view it on GitHub .

-- 
Matt Benjamin
CohortFS, LLC.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://cohortfs.com

tel.  734-761-4689 
fax.  734-769-8938 
cel.  734-216-5309

File: src/msg/xio/XioMessenger.cc [hunk 1]
```diff
@@ -0,0 +1,1041 @@
...
+      /* enable flow-control */
+      xopt = 1;
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_ENABLE_FLOW_CONTROL,
+                 &xopt, sizeof(xopt));
+
+      /* and set threshold for buffer callouts */
+      xopt = 16384;
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_MAX_INLINE_DATA,
+                 &xopt, sizeof(xopt));
+      xopt = 216;
```
Comment (rajukv) [inline src/msg/xio/XioMessenger.cc]: Matt, can you increase the max inline header size supported to atleast 218 bytes? Kernel RBD sends the ceph header as is(without encoding) and contains a 2 byte reserved field that is not handled if max inline header size in just 216.

File: src/msg/xio/XioMsg.h [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+    ::encode(hdr->type, bl);
+    ::encode(hdr->priority, bl);
+    ::encode(hdr->version, bl);
+    ::encode(hdr->front_len, bl);
+    ::encode(hdr->middle_len, bl);
+    ::encode(hdr->data_len, bl);
+    ::encode(hdr->data_off, bl);
+    ::encode(hdr->src.type, bl);
+    ::encode(hdr->src.num, bl);
+    ::encode(hdr->compat_version, bl);
```
Comment (rajukv) [inline src/msg/xio/XioMsg.h]: Can you add 
::encode(hdr->reserved, bl)
after line 84? Kernel RBD can send the incore ceph_msg_hdr without actually doing an encode for each request(as is done today).

File: src/msg/xio/XioMsg.h [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+    ::decode(hdr->type, bl);
+    ::decode(hdr->priority, bl);
+    ::decode(hdr->version, bl);
+    ::decode(hdr->front_len, bl);
+    ::decode(hdr->middle_len, bl);
+    ::decode(hdr->data_len, bl);
+    ::decode(hdr->data_off, bl);
+    ::decode(hdr->src.type, bl);
+    ::decode(hdr->src.num, bl);
+    ::decode(hdr->compat_version, bl);
```
Comment (rajukv) [inline src/msg/xio/XioMsg.h]: Can you add 
::decode(hdr->reserved, bl)
after line 116? Same reason as above

**Comment by mattbenjamin** [comment]
The 216 actually is more than the encoded or (I suspect) in-core size of the ceph_header, so I think this is a non-problem that will be fixed when CephX/authorizer work is added.  So I don't think this poses a problem for Linux kernel interop.  Thanks for calling attention to the header, though, it hasn't been discussed much.

**Comment by mattbenjamin** [comment]
I eyeskipped the reserved word;  I mean to say above, the current wire header can still be changed, but it's not critical to change at this patch set, which is experimental;  in addition to the point that the 216 byte value is misleading.

**Comment by liewegas** [comment]
merged manually!

**Pull request closed without merging**